server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java

```diff
@@ -320,22 +322,27 @@ private void sendNodeRequest(final DiscoveryNode node, List<ShardRouting> shards
                 if (task != null) {
                     nodeRequest.setParentTask(clusterService.localNode().getId(), task.getId());
                 }
-                transportService.sendRequest(node, transportNodeBroadcastAction, nodeRequest, new TransportResponseHandler<NodeResponse>() {
-                    @Override
-                    public NodeResponse read(StreamInput in) throws IOException {
-                        return new NodeResponse(in);
-                    }
-
-                    @Override
-                    public void handleResponse(NodeResponse response) {
-                        onNodeResponse(node, nodeIndex, response);
-                    }
 
-                    @Override
-                    public void handleException(TransportException exp) {
-                        onNodeFailure(node, nodeIndex, exp);
-                    }
-                });
+                final TransportRequestOptions transportRequestOptions = request.timeout().equals(TimeValue.MINUS_ONE)
+                        ? TransportRequestOptions.EMPTY : TransportRequestOptions.timeout(request.timeout());
```

Comment:
Not super happy we have to special case this here. I wonder if we should either make the `request.timeout()` field Nullable (and always pass it on here), or if we should special case the `-1` in TransportService not to create a `TimeoutHandler`. What's even weirder is that `ScheduledThreadPoolExecutor` is treating negative values as instant trigger.

Reply:
Sure, see a451bad.